src/lib.rs

```diff
@@ -16,88 +15,184 @@ pub fn run<F>(f: F)
 where
     F: Future<Output = ()> + Send + 'static,
 {
-    namespace::unshare_user().unwrap();
+    unshare_user().unwrap();
     async_global_executor::block_on(f);
 }
 
-#[derive(Debug)]
-pub struct Machine<C, E> {
-    id: u64,
-    addr: Ipv4Addr,
-    mask: u8,
-    ctrl: mpsc::Sender<IfaceCtrl>,
-    tx: mpsc::UnboundedSender<C>,
-    rx: mpsc::UnboundedReceiver<E>,
+enum Connector {
+    Unplugged(Plug),
+    Plugged(NetworkId),
+    Shutdown,
 }
 
-impl<C: Send + 'static, E: Send + 'static> Machine<C, E> {
-    pub fn id(&self) -> u64 {
-        self.id
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct NetworkId(usize);
+
+impl NetworkId {
+    fn id(&self) -> usize {
+        self.0 + u16::MAX as usize
     }
+}
 
-    pub fn addr(&self) -> Ipv4Addr {
-        self.addr
+pub struct Netsim<C, E> {
+    machines: Vec<Machine<C, E>>,
+    plugs: Vec<Connector>,
+    networks: Vec<Network>,
+}
+
+impl<C, E> Default for Netsim<C, E> {
```

Comment:
replace with `#[derive(Default)]`?

Reply:
That requires a Default trait bound on C and E which are probably going to be enums